src/canisters/frontend/ic-asset/src/asset_config.rs

```diff
@@ -100,8 +134,28 @@ impl AssetSourceDirectoryConfiguration {
                     parent_dir
                 )
             })?
-            .get_config(canonical_path))
+            .lock().unwrap().get_config(canonical_path))
     }
+
+    pub(crate) fn get_unused_configs(&self) -> HashSet<AssetConfigRule> {
```

Comment:
needed to dedup the list, because same `AssetConfigRule` gets referenced in parent dir as well as all its subdirs, so e.g. if you have `.ic-assets.json` in `/.ic-assets.json` and rule `X` in it, and you have `/dir` and `/dir/dir`, then the rule `X` will be also present for `/dir` and `/dir/dir` directories (so total of 3 copies)

I decided to dedup stuff with `HashSet` - better runtime performance at the cost of 
- code clarity (needed to `impl`  `Hash` and `Eq/PartialEq` for several structs) 
- memory (to build `HashSet` this is not significant)

Reply:
Not sure I follow.  I thought `/dir/dir` would first process rules in `/dir`, then process its own.  So no dupes.  Did something change?

Reply:
good catch, the reason rules get duplicated is because of this:
https://github.com/dfinity/sdk/blob/d5f9be9608ab174e0c66405babed45c19c2612ec/src/canisters/frontend/ic-asset/src/asset_config.rs#L265-L268
, but when changed to: 
```rust
        let parent_ref = match parent {
            // Some(p) if rules.is_empty() => p,
            _ => Arc::new(Mutex::new(Self { parent, rules })),
        };
```
I no longer get dups, and all unit+e2e tests are passing. 

Any chance you remember why this was added? 🙄

Reply:
currently on `master`, when`dbg!(&configuration)` is added on line 126 in `sync.rs`:
https://github.com/dfinity/sdk/blob/db3de4defbeb5a171152ee366fc51e52260804b4/src/canisters/frontend/ic-asset/src/sync.rs#L67-L128 and we would `cargo test -- --nocapture` to see `dbg!` output from tests,  this would be the output from `#[test] fn gather_all_nondot_files_from_nondot_dirs()` (picked at random):
```rust
[src/canisters/frontend/ic-asset/src/sync.rs:126] &configuration = AssetSourceDirectoryConfiguration {
    config_map: {
        "/private/var/folders/qr/2sz_4fyx75z7smnl9k554p000000gn/T/assetsuGbeN/.hidden-dir-flat": AssetConfigTreeNode {
            parent: None,
            rules: [
                AssetConfigRule {
                    r#match: /private/var/folders/qr/2sz_4fyx75z7smnl9k554p000000gn/T/assetsuGbeN/**/*,
                    cache: None,
                    headers: Absent,
                    ignore: None,
                },
            ],
        },
        "/private/var/folders/qr/2sz_4fyx75z7smnl9k554p000000gn/T/assetsuGbeN/.hidden-dir/.hidden-dir-nested": AssetConfigTreeNode {
            parent: None,
            rules: [
                AssetConfigRule {
                    r#match: /private/var/folders/qr/2sz_4fyx75z7smnl9k554p000000gn/T/assetsuGbeN/**/*,
                    cache: None,
                    headers: Absent,
                    ignore: None,
                },
            ],
        },
        "/private/var/folders/qr/2sz_4fyx75z7smnl9k554p000000gn/T/assetsuGbeN/.hidden-dir": AssetConfigTreeNode {
            parent: None,
            rules: [
                AssetConfigRule {
                    r#match: /private/var/folders/qr/2sz_4fyx75z7smnl9k554p000000gn/T/assetsuGbeN/**/*,
                    cache: None,
                    headers: Absent,
                    ignore: None,
                },
            ],
        },
        "/private/var/folders/qr/2sz_4fyx75z7smnl9k554p000000gn/T/assetsuGbeN": AssetConfigTreeNode {
            parent: None,
            rules: [
                AssetConfigRule {
                    r#match: /private/var/folders/qr/2sz_4fyx75z7smnl9k554p000000gn/T/assetsuGbeN/**/*,
                    cache: None,
                    headers: Absent,
                    ignore: None,
                },
            ],
        },
    },
}
```
the following would get printed after we comment `            // Some(p) if rules.is_empty() => p,`:
```rust
[src/canisters/frontend/ic-asset/src/sync.rs:126] &configuration = AssetSourceDirectoryConfiguration {
    config_map: {
        "/private/var/folders/qr/2sz_4fyx75z7smnl9k554p000000gn/T/assetsiWZ3k/.hidden-dir-flat": AssetConfigTreeNode {
            parent: Some(
                AssetConfigTreeNode {
                    parent: None,
                    rules: [
                        AssetConfigRule {
                            r#match: /private/var/folders/qr/2sz_4fyx75z7smnl9k554p000000gn/T/assetsiWZ3k/**/*,
                            cache: None,
                            headers: Absent,
                            ignore: None,
                        },
                    ],
                },
            ),
            rules: [],
        },
        "/private/var/folders/qr/2sz_4fyx75z7smnl9k554p000000gn/T/assetsiWZ3k/.hidden-dir/.hidden-dir-nested": AssetConfigTreeNode {
            parent: Some(
                AssetConfigTreeNode {
                    parent: Some(
                        AssetConfigTreeNode {
                            parent: None,
                            rules: [
                                AssetConfigRule {
                                    r#match: /private/var/folders/qr/2sz_4fyx75z7smnl9k554p000000gn/T/assetsiWZ3k/**/*,
                                    cache: None,
                                    headers: Absent,
                                    ignore: None,
                                },
                            ],
                        },
                    ),
                    rules: [],
                },
            ),
            rules: [],
        },
        "/private/var/folders/qr/2sz_4fyx75z7smnl9k554p000000gn/T/assetsiWZ3k/.hidden-dir": AssetConfigTreeNode {
            parent: Some(
                AssetConfigTreeNode {
                    parent: None,
                    rules: [
                        AssetConfigRule {
                            r#match: /private/var/folders/qr/2sz_4fyx75z7smnl9k554p000000gn/T/assetsiWZ3k/**/*,
                            cache: None,
                            headers: Absent,
                            ignore: None,
                        },
                    ],
                },
            ),
            rules: [],
        },
        "/private/var/folders/qr/2sz_4fyx75z7smnl9k554p000000gn/T/assetsiWZ3k": AssetConfigTreeNode {
            parent: None,
            rules: [
                AssetConfigRule {
                    r#match: /private/var/folders/qr/2sz_4fyx75z7smnl9k554p000000gn/T/assetsiWZ3k/**/*,
                    cache: None,
                    headers: Absent,
                    ignore: None,
                },
            ],
        },
    },
}
```
which does seem like the thing we wanted to have originally...

Reply:
found the original suggestion, see: https://github.com/dfinity/agent-rs/pull/352#discussion_r912211218

Reply:
I'm gonna get rid of this tree-pruning feature unless you have a better idea?